## FE520 Assignment 5 JIALE QIN

## Data Processing with Pandas

In [1]:
import numpy as np
import pandas as pd
from pandas import DataFrame

#### 1. Read 'Energy.xlsx' and 'EnergyRating.xlsx' as BalanceSheet and Ratings (dataframe)

In [2]:
BalanceSheet = pd.read_excel('Energy.xlsx')
BalanceSheet.head()

Global Company Key  Data Date  Fiscal Year  Fiscal Quarter  \
0                1380   20100331         2010               1   
1                1380   20100630         2010               2   
2                1380   20100930         2010               3   
3                1380   20101231         2010               4   
4                1380   20110331         2011               1   

   Fiscal Year-end Month Industry Format  \
0                     12            INDL   
1                     12            INDL   
2                     12            INDL   
3                     12            INDL   
4                     12            INDL   

  Level of Consolidation - Company Interim Descriptor Population Source  \
0                                                  C                  D   
1                                                  C                  D   
2                                                  C                  D   
3                                                  C                  D   
4                                                  C                  D   

  Data Format Ticker Symbol                    ...                     \
0         STD           HES                    ...                      
1         STD           HES                    ...                      
2         STD           HES                    ...                      
3         STD           HES                    ...                      
4         STD           HES                    ...                      

  Implied Option EPS Basic Implied Option EPS Basic Preliminary  \
0                      0.0                                  0.0   
1                      0.0                                  0.0   
2                      0.0                                  0.0   
3                      0.0                                  0.0   
4                      0.0                                  0.0   

  Implied Option Expense Implied Option Expense Preliminary  \
0                    0.0                                0.0   
1                    0.0                                0.0   
2                    0.0                                0.0   
3                    0.0                                0.0   
4                    0.0                                0.0   

  Research and Development Expense  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

   Selling, General and Administrative Expenses  Stock Exchange Code  \
0                                         559.0                   11   
1                                         576.0                   11   
2                                         608.0                   11   
3                                         805.0                   11   
4                                         760.0                   11   

   CIK Number  Active/Inactive Status Marker  \
0        4447                              A   
1        4447                              A   
2        4447                              A   
3        4447                              A   
4        4447                              A   

   Current ISO Country Code - Incorporation  
0                                       USA  
1                                       USA  
2                                       USA  
3                                       USA  
4                                       USA  

[5 rows x 380 columns]

In [3]:
Ratings = pd.read_excel('EnergyRating.xlsx')
Ratings.head()

Global Company Key S&P Domestic Long Term Issuer Credit Rating  \
0                1380                                        BBB-   
1                1380                                        BBB-   
2                1380                                        BBB-   
3                1380                                        BBB-   
4                1380                                        BBB-   

   S&P Subordinated Debt Rating S&P Domestic Short Term Issuer Credit Rating  \
0                           NaN                                          NaN   
1                           NaN                                          NaN   
2                           NaN                                          NaN   
3                           NaN                                          NaN   
4                           NaN                                          NaN   

   Data Date                           Address Line 1 Ticker Symbol  
0   20100131  1185 Avenue of the Americas, 40th Floor           HES  
1   20100228  1185 Avenue of the Americas, 40th Floor           HES  
2   20100331  1185 Avenue of the Americas, 40th Floor           HES  
3   20100430  1185 Avenue of the Americas, 40th Floor           HES  
4   20100531  1185 Avenue of the Americas, 40th Floor           HES

#### 2. Replace all None or NaN with 0

In [4]:
BalanceSheet = BalanceSheet.fillna(0)
# Make sure there is no nan
np.any(BalanceSheet.isnull()) == True

False

In [5]:
Ratings = Ratings.fillna(0)
# Make sure there is no nan
np.any(Ratings.isnull()) == True

False

#### 3. Drop the column if more than 90% value in this column is 0. (Hint, using pd.apply() function)

In [6]:
BalanceSheet = BalanceSheet.drop(BalanceSheet.columns[((BalanceSheet==0).mean()>0.9)], axis=1)

In [7]:
BalanceSheet.shape

(844, 234)

The column of BalanceSheet reduced to 234.

#### 4. Normalize the table (columns from Accumulated Other Comprehensive Income (Loss) to Selling, General and Administrative Expenses)

Using pd.apply() to normalize the table, in this table, you need to calculate the largest value, and smallest value in each column, compute the range for each variable. Then using value in this column divided by range of this column. (Do not using any function like MinMax(), you need to write it by yourself)

In [8]:
def normalize(column):
    # Find the max and min value
    max_x = column[0]
    min_x = column[0]
    for item in column:
        if item < min_x:
            min_x = item
        if item > max_x:
            max_x = item
    
    # Calculate the normalized value and return transformed values
    for i in range(len(column)):
        column[i] = column[i]/(max_x - min_x)
    
    return column

In [9]:
BalanceSheet.loc[:,'Accumulated Other Comprehensive Income (Loss)':'Selling, General and Administrative Expenses'] = BalanceSheet.loc[:,'Accumulated Other Comprehensive Income (Loss)':'Selling, General and Administrative Expenses'].apply(normalize)

In [10]:
BalanceSheet.head()

Global Company Key  Data Date  Fiscal Year  Fiscal Quarter  \
0                1380   20100331         2010               1   
1                1380   20100630         2010               2   
2                1380   20100930         2010               3   
3                1380   20101231         2010               4   
4                1380   20110331         2011               1   

   Fiscal Year-end Month Industry Format  \
0                     12            INDL   
1                     12            INDL   
2                     12            INDL   
3                     12            INDL   
4                     12            INDL   

  Level of Consolidation - Company Interim Descriptor Population Source  \
0                                                  C                  D   
1                                                  C                  D   
2                                                  C                  D   
3                                                  C                  D   
4                                                  C                  D   

  Data Format Ticker Symbol                    ...                     \
0         STD           HES                    ...                      
1         STD           HES                    ...                      
2         STD           HES                    ...                      
3         STD           HES                    ...                      
4         STD           HES                    ...                      

  Accrued Expenses Extraordinary Items and Discontinued Operations  \
0              0.0                                             0.0   
1              0.0                                             0.0   
2              0.0                                             0.0   
3              0.0                                             0.0   
4              0.0                                             0.0   

  Interest and Related Expense- Total Operating Expense- Total  \
0                            0.096481                 0.058246   
1                            0.095346                 0.048961   
2                            0.107832                 0.049267   
3                            0.115778                 0.058033   
4                            0.114642                 0.063935   

  Research and Development Expense  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   

   Selling, General and Administrative Expenses  Stock Exchange Code  \
0                                      0.124443                   11   
1                                      0.128228                   11   
2                                      0.135352                   11   
3                                      0.179207                   11   
4                                      0.169190                   11   

   CIK Number  Active/Inactive Status Marker  \
0        4447                              A   
1        4447                              A   
2        4447                              A   
3        4447                              A   
4        4447                              A   

   Current ISO Country Code - Incorporation  
0                                       USA  
1                                       USA  
2                                       USA  
3                                       USA  
4                                       USA  

[5 rows x 234 columns]

#### 5. Calculate the correlation matrix for variables = ['Current Assets - Other - Total', 'Current Assets - Total', 'Other Long-term Assets', 'Assets Netting Other Adjustments']

In [12]:
correlation = BalanceSheet[['Current Assets - Other - Total','Current Assets - Total', 'Other Long-term Assets', 'Assets Netting & Other Adjustments']]

In [13]:
correlation = correlation.corr()
correlation

Current Assets - Other - Total  \
Current Assets - Other - Total                            1.000000   
Current Assets - Total                                    0.791628   
Other Long-term Assets                                    0.642217   
Assets Netting & Other Adjustments                        0.056326   

                                    Current Assets - Total  \
Current Assets - Other - Total                    0.791628   
Current Assets - Total                            1.000000   
Other Long-term Assets                            0.689530   
Assets Netting & Other Adjustments               -0.057167   

                                    Other Long-term Assets  \
Current Assets - Other - Total                    0.642217   
Current Assets - Total                            0.689530   
Other Long-term Assets                            1.000000   
Assets Netting & Other Adjustments               -0.005107   

                                    Assets Netting & Other Adjustments  
Current Assets - Other - Total                                0.056326  
Current Assets - Total                                       -0.057167  
Other Long-term Assets                                       -0.005107  
Assets Netting & Other Adjustments                            1.000000

#### 6. If you look at column ('Company Name'), you will find some company name end with 'CORP', 'CO' or 'INC'. Create a new column (Name: 'CO') to store the last word of company name. (For example: 'CORP', 'CO' or 'INC') (Hint: using map function)

In [14]:
BalanceSheet['CO'] = BalanceSheet['Company Name'].str.split(" ").apply(lambda x: x[-1])

#### 7. Merge (inner) Ratings and BalanceSheet based on 'datadate' and 'Global Company Key', and name merged dataset 'Matched'.

In [15]:
Matched = pd.merge(BalanceSheet, Ratings, on = ['Data Date', 'Global Company Key'], how = 'inner')

In [16]:
Matched.shape

(822, 240)

#### 8. Mapping

For dataset 'Matched', we have following mapping:  
AAA = 0  
AA+ = 1  
AA = 2  
AA- = 3  
A+ = 4  
A = 5   
A- = 6  
BBB+ = 7  
BBB = 8  
BBB- = 9  
BB+ = 10  
BB = 11  
others = 12  
Using map function to create a new variable = 'Rate', which maps ratings to numerical ratings.

In [17]:
Rate = {'AAA': 0,
'AA+': 1,
'AA' : 2,
'AA-' : 3,
'A+' : 4,
'A' : 5,
'A-' : 6,
'BBB+' : 7,
'BBB' : 8,
'BBB-' : 9,
'BB+' : 10,
'BB' : 11,
'others' : 12}

In [18]:
Matched['Rate'] = Matched['S&P Domestic Long Term Issuer Credit Rating'].map(Rate)

In [19]:
Matched['Rate'].head()

0    9.0
1    9.0
2    8.0
3    8.0
4    8.0
Name: Rate, dtype: float64

#### 9. Calculate the rating frequency of company whose name end with 'CO'

In [20]:
# Select column Rate and CO
company = Matched[['Rate', 'CO']]
company.head()

Rate    CO
0   9.0  CORP
1   9.0  CORP
2   8.0  CORP
3   8.0  CORP
4   8.0  CORP

In [21]:
# Select the company whose name end with 'CO'
company = company[company['CO'] == 'CO']
company

Rate  CO
140   5.0  CO
141   5.0  CO
142   5.0  CO
143   5.0  CO
144   5.0  CO
145   5.0  CO
146   5.0  CO
147   5.0  CO
148   5.0  CO
149   5.0  CO
150   5.0  CO
151   5.0  CO
152   5.0  CO
153   5.0  CO
154   5.0  CO
155   5.0  CO
156   5.0  CO
157   5.0  CO
158   5.0  CO
159   5.0  CO
160   5.0  CO
161   5.0  CO
162   5.0  CO
163   5.0  CO
164   5.0  CO
165   6.0  CO
166   7.0  CO
167   7.0  CO
504  10.0  CO
505  10.0  CO
..    ...  ..
664  10.0  CO
665  10.0  CO
722  11.0  CO
723  11.0  CO
724  11.0  CO
725  11.0  CO
726  11.0  CO
727  10.0  CO
728  10.0  CO
729  10.0  CO
730  10.0  CO
731  10.0  CO
732  10.0  CO
733  10.0  CO
734  10.0  CO
735  10.0  CO
736  10.0  CO
737  10.0  CO
738  10.0  CO
739  10.0  CO
740  10.0  CO
741  10.0  CO
742  10.0  CO
743   9.0  CO
744   9.0  CO
745   9.0  CO
746   9.0  CO
747   9.0  CO
748   9.0  CO
749   9.0  CO

[112 rows x 2 columns]

In [22]:
company.groupby('Rate').count()

CO
Rate    
5.0   25
6.0    1
7.0    2
9.0   50
10.0  29
11.0   5

#### Outpyt your final dataset as 'HW4.csv'

In [23]:
Matched.to_csv('HW4.csv')